### ISSUE TITLE
Cancel after consume( ... { no_ack => 1 }) results in SEGV

### ISSUE THREAD

**Issue by giganteous** [net-amqp-rabbitmq/net-amqp-rabbitmq#97]
Hey,

Whilst on a mission for some easy examples in perl and python (using an acknowledgement per frame), I ran into a Segfault.

The code:
```perl
#!/usr/bin/env perl
use strict;
use warnings;
use Net::AMQP::RabbitMQ;

my $mq = Net::AMQP::RabbitMQ->new();
$mq->connect('localhost', {
        user => $ENV{"MQUSER"},
        password => $ENV{"MQPASS"},
        });
$mq->channel_open(1);
my $tag = $mq->consume(1, "aqueue"); # no_ack defaults to 1
my $count = 0;
while ( my $frame = $mq->recv(0) ) {
        print $frame->{body} . "\n";
        $mq->ack(CHAN, $frame->{delivery_tag});
        $count ++;
        if ($count >= 10 ) {
            my $res = $mq->cancel(CHAN, $tag);
            print "Cancel result: $res\n";
            last;
        }
}
$mq->disconnect();
```

Code to fix it:
```patch
Index: build-area/RabbitMQ.xs
===================================================================
--- build-area.orig/RabbitMQ.xs
+++ build-area/RabbitMQ.xs
@@ -1378,6 +1378,11 @@ net_amqp_rabbitmq_cancel(conn, channel,
     assert_amqp_connected(conn);
 
     r = amqp_basic_cancel(conn, channel, amqp_cstring_bytes(consumer_tag));
+    if(r == NULL)
+    {
+      Perl_croak(aTHX_ "Basic cancel failed");
+    }
+
     if(strlen(consumer_tag) == r->consumer_tag.len && 0 == strcmp(consumer_tag, (char *)r->consumer_tag.bytes)) {
       RETVAL = 1;
     } else {
```

But I suppose there might be some handy info: telling the user why it's not cool to call cancel.

Cheers

**Comment by giganteous** [comment]
Copy and pasting a bit, I came up with adding this:

```patch
--- RabbitMQ.xs.orig	2016-03-09 22:35:14.187389443 +0100
+++ RabbitMQ.xs	2016-03-09 22:29:31.187393846 +0100
@@ -1378,6 +1378,7 @@
     assert_amqp_connected(conn);
 
     r = amqp_basic_cancel(conn, channel, amqp_cstring_bytes(consumer_tag));
+    die_on_amqp_error(aTHX_ amqp_get_rpc_reply(conn), conn, "cancel");
     if(strlen(consumer_tag) == r->consumer_tag.len && 0 == strcmp(consumer_tag, (char *)r->consumer_tag.bytes)) {
       RETVAL = 1;
     } else {
```

gives me this error:

```
cancel: server channel error 406, message: PRECONDITION_FAILED - unknown delivery tag 1 at consume_queue.pl line 26.
```

I'm not at all into .xs. Is this any good?

Cheers

**Comment by fgabolde** [comment]
I got burned by this one too.

I don't understand the error message you got ("unknown delivery tag 1") but it would help to have an exception thrown instead of segfaulting, could this patch be merged?

**Comment by markwellis** [comment]
Sorry I didn't see this. I'll look at it this week

**Comment by manchicken** [comment]
You can just add a new macro for channels like we have for the macro `assert_amqp_connected()`. That seems like it would be the best way to handle that.

**Comment by markwellis** [comment]
hmm, this seems to be one of those weird bugs caused by ack being async, and cancel not checking the current rpc status.

so the ack failed, because "unknown delivery tag 1". this leaves the protocol in a weird state, so the cancel fails. but we don't check if the cancel failed, so we do r->consumer_tag, where r is NULL and segfault.

I think adding the `die_on_amqp_error(aTHX_ amqp_get_rpc_reply(conn), conn, "cancel");` and a check to make sure r isn't NULL is the best fix. But i've not wrote c in a while so maybe not.

@manchicken what do you think? also I don't get what you mean about added a channel macro like `assert_amqp_connected`

**Comment by manchicken** [comment]
I wonder if it was a connection failure in the middle of the loop, too. We might want to just detect a closed channel and see if that helps.

**Comment by markwellis** [comment]
Good plan, I'll try that tonight. Although my test is a little different, as I have no loop. As long as you ack on something that should error then call cancel it segfaults.

**Comment by manchicken** [comment]
Ooh! Both? We should do both.

**Comment by giganteous** [comment]
@manchicken you ask if "it" was a connection failure. Not in my case; it always segfaults.

Thanks for picking it up!

**Comment by giganteous** [comment]
The puzzle stats to slowly come together again.

Isn't my cancel a misuse of rabbitmq? When running the same code with this:

> my $tag = $mq->consume(CHAN, "otp", { no_ack => 0 });

the cancel runs fine. It is improper use of cancel IIUC. But in any case, the cancel shouldn't give a segfault.

Hence my question: What bone to throw to the developer: A cancel on a NO_ACK==1 tag should give a friendly critical/error that its not proper use of channels.

I'm no rabbitmq expert though.

**Comment by markwellis** [comment]
It's not the cancel that's failing for me. It's the ack on an invalid delivery tag. That causes a protocol error that causes the cancel to fail. We're not checking if it fails, so it segfaults.

I've not tried using cancel properly yet, I'd assume that gives an error, there's some code around that.

In your test, you have `$mq->ack(CHAN, $frame->{delivery_tag});` but CHAN is not defined, so I'm not sure how that's working as you have strict and warnings. I can only assume you've either added the use strict in for this comment, which would explain the segfault (ack on an invalid consumer/delivery tag), CHAN is set to something invalid or it's just an error in copy pasting.

If you could post the actual code that segfaults that would be great, just to make sure I catch all the corner cases :)

**Comment by fgabolde** [comment]
@markwellis I have a minimal example that successfully segfaults. To make it not segfault, just switch to `no_ack => 0`, or don't call `ack()`.

```perl
#!/usr/bin/env perl

use strict;
use warnings;
use 5.012;
use Carp;
use autodie;
use utf8;

use Net::AMQP::RabbitMQ;

my $mq = Net::AMQP::RabbitMQ->new();
$mq->connect('localhost',
             { port => 5672,
               user => 'guest',
               password => 'guest' });

my $channel = 1;
$mq->channel_open($channel);
my $consumer_tag = $mq->consume($channel, 'foo', { no_ack => 1 });
my $message = $mq->recv(0);

use Data::Dumper;
print Dumper($message); # looks good

$mq->ack($channel, $message->{delivery_tag});
$mq->cancel($channel, $consumer_tag); # segfaults here

exit 0;
```

**Comment by markwellis** [comment]
Cool thanks :)

**Comment by giganteous** [comment]
@markwellis The undefined CHAN is the only typo in my script, set it to 1 to see the observed behaviour. I tried to be helpful by cutting away some cruft. I should remember to not do that! 

If there's less than 10 in the queue, the behaviour is really weird: After the first ack() it'll hang forever, so it really needs a no_ack => 0 to be an example.

**Comment by manchicken** [comment]
@markwellis, I wonder if we need to have something set an alarm and use some manner of flutex or something to cause a wait loop on the cancel? It might make sense to use a mutex/flutex for the channel for some of this stuff.

Ruby's library uses a mutex, https://github.com/ruby-amqp/amqp/blob/5d6102bfb4155c8ecc9263267eff5cd0cc791713/lib/amqp/channel.rb

It kinda makes sense to start doing that here.

**Comment by markwellis** [comment]
@manchicken I don't follow.

The error in this case (the segfault) is because the cancel failed because the ack failed, and we never check if the cancel result is NULL before we dereference it.

As I understand it, some parts of AMQP are async by design, and ack is one of them. The way we normally handle this is to check for a RPC error after calling a synchronous function, which would error if the previous async function failed. At that point, we die so don't run into null dereference issues.

I'm not sure what the mutex is for in the ruby library, I had a quick look and it seemed to be to make async functions sync? But that doesn't make sense to me.

So this is just a case of not error checking the amqp_basic_cancel result and it failing due to reasons which causes a segfault. The ack part of it is just a way to make it fail reliably..

I could be wrong though, but that's how I currently understand it. But I've spent all of 10 mins looking at this so I could be way off.

**Comment by markwellis** [comment]
So the correct fix is a combination of the two patches originally suggested

```patch
Index: build-area/RabbitMQ.xs
===================================================================
--- build-area.orig/RabbitMQ.xs
+++ build-area/RabbitMQ.xs
@@ -1378,6 +1378,11 @@ net_amqp_rabbitmq_cancel(conn, channel,
     assert_amqp_connected(conn);
 
     r = amqp_basic_cancel(conn, channel, amqp_cstring_bytes(consumer_tag));
+    if(r == NULL)
+    {
+      Perl_croak(aTHX_ "Basic cancel failed");
+    }
+
     if(strlen(consumer_tag) == r->consumer_tag.len && 0 == strcmp(consumer_tag, (char *)r->consumer_tag.bytes)) {
       RETVAL = 1;
     } else {
```

to check if the cancel failed, and

```patch
--- RabbitMQ.xs.orig	2016-03-09 22:35:14.187389443 +0100
+++ RabbitMQ.xs	2016-03-09 22:29:31.187393846 +0100
@@ -1378,6 +1378,7 @@
     assert_amqp_connected(conn);
 
     r = amqp_basic_cancel(conn, channel, amqp_cstring_bytes(consumer_tag));
+    die_on_amqp_error(aTHX_ amqp_get_rpc_reply(conn), conn, "cancel");
     if(strlen(consumer_tag) == r->consumer_tag.len && 0 == strcmp(consumer_tag, (char *)r->consumer_tag.bytes)) {
       RETVAL = 1;
     } else {
```

to report the RPC failure reason (which might not be set, the cancel could fail for any number of reasons)

**Comment by markwellis** [comment]
Fixed in version 2.30000, which will be on cpan shortly
